File: snobot_sim_java/src/main/java/com/snobot/simulator/simulator_components/ctre/CtreManager.java
package com.snobot.simulator.simulator_components.ctre;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

import com.snobot.simulator.simulator_components.smart_sc.BaseCanSmartSpeedController;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import com.snobot.simulator.SensorActuatorRegistry;
import com.snobot.simulator.wrapper_accessors.DataAccessorFactory;

@SuppressWarnings({"PMD.NcssCount", "PMD.CyclomaticComplexity"})
public class CtreManager
{
    private static final Logger sLOGGER = LogManager.getLogger(CtreManager.class);

    private final Map<Integer, CtrePigeonImuSim> mPigeonMap;

    private final Map<String, Consumer<CtreTalonSrxSpeedControllerSim>> mMotorControllerNormalCallbacks;
    private final Map<String, BiConsumer<CtreTalonSrxSpeedControllerSim, Integer>> mMotorControllerSlottedCallbacks;

    private final Map<String, Consumer<CtrePigeonImuSim>> mPigeonNormalCallbacks;

    public CtreManager()
    {
        mPigeonMap = new HashMap<>();
        
        mMotorControllerNormalCallbacks = new HashMap<>();
        mMotorControllerNormalCallbacks.put("SetDemand", CtreTalonSrxSpeedControllerSim::handleSetDemand);
        mMotorControllerNormalCallbacks.put("Set_4", CtreTalonSrxSpeedControllerSim::handleSet4);
        mMotorControllerNormalCallbacks.put("ConfigSelectedFeedbackSensor", CtreTalonSrxSpeedControllerSim::handleConfigSelectedFeedbackSensor);
        mMotorControllerNormalCallbacks.put("ConfigMotionCruiseVelocity", CtreTalonSrxSpeedControllerSim::handleSetMotionCruiseVelocity);
        mMotorControllerNormalCallbacks.put("ConfigMotionAcceleration", CtreTalonSrxSpeedControllerSim::handleSetMotionAcceleration);
        mMotorControllerNormalCallbacks.put("PushMotionProfileTrajectory", CtreTalonSrxSpeedControllerSim::handlePushMotionProfileTrajectory);
        mMotorControllerNormalCallbacks.put("PushMotionProfileTrajectory_2", CtreTalonSrxSpeedControllerSim::handlePushMotionProfileTrajectory_2);
        mMotorControllerNormalCallbacks.put("SetSelectedSensorPosition", CtreTalonSrxSpeedControllerSim::handleSetSelectedSensorPosition);
        mMotorControllerNormalCallbacks.put("SetInverted_2", CtreTalonSrxSpeedControllerSim::handleSetInverted_2);

        mMotorControllerNormalCallbacks.put("GetMotorOutputPercent", CtreTalonSrxSpeedControllerSim::handleGetMotorOutputPercent);
        mMotorControllerNormalCallbacks.put("GetSelectedSensorPosition", CtreTalonSrxSpeedControllerSim::handleGetSelectedSensorPosition);
        mMotorControllerNormalCallbacks.put("GetSelectedSensorVelocity", CtreTalonSrxSpeedControllerSim::handleGetSelectedSensorVelocity);
        mMotorControllerNormalCallbacks.put("GetClosedLoopError", CtreTalonSrxSpeedControllerSim::handleGetClosedLoopError);
        mMotorControllerNormalCallbacks.put("GetMotionProfileStatus", CtreTalonSrxSpeedControllerSim::handleGetMotionProfileStatus);
        mMotorControllerNormalCallbacks.put("GetActiveTrajectoryPosition", CtreTalonSrxSpeedControllerSim::handleGetActiveTrajectoryPosition);
        mMotorControllerNormalCallbacks.put("GetActiveTrajectoryVelocity", CtreTalonSrxSpeedControllerSim::handleGetActiveTrajectoryVelocity);
        mMotorControllerNormalCallbacks.put("GetQuadraturePosition", CtreTalonSrxSpeedControllerSim::handleGetQuadraturePosition);
        mMotorControllerNormalCallbacks.put("GetQuadratureVelocity", CtreTalonSrxSpeedControllerSim::handleGetQuadratureVelocity);

        mMotorControllerSlottedCallbacks = new HashMap<>();
        mMotorControllerSlottedCallbacks.put("Config_kP", CtreTalonSrxSpeedControllerSim::handleSetKP);
        mMotorControllerSlottedCallbacks.put("Config_kI", CtreTalonSrxSpeedControllerSim::handleSetKI);
        mMotorControllerSlottedCallbacks.put("Config_kD", CtreTalonSrxSpeedControllerSim::handleSetKD);
        mMotorControllerSlottedCallbacks.put("Config_kF", CtreTalonSrxSpeedControllerSim::handleSetKF);
        mMotorControllerSlottedCallbacks.put("Config_IntegralZone", CtreTalonSrxSpeedControllerSim::handleSetIntegralZone);

        mPigeonNormalCallbacks = new HashMap<>();
        mPigeonNormalCallbacks.put("SetYaw", CtrePigeonImuSim::handleSetYaw);
        mPigeonNormalCallbacks.put("GetRawGyro", CtrePigeonImuSim::handleGetRawGyro);
        mPigeonNormalCallbacks.put("GetYawPitchRoll", CtrePigeonImuSim::handleGetYawPitchRoll);
        mPigeonNormalCallbacks.put("GetFusedHeading", CtrePigeonImuSim::handleGetFusedHeading);
        mPigeonNormalCallbacks.put("GetFusedHeading1", CtrePigeonImuSim::handleGetFusedHeading);
    }

    public void reset()
    {
        mPigeonMap.clear();
    }

    private CtreTalonSrxSpeedControllerSim getMotorControllerWrapper(int aCanPort)
    {
        return CtreTalonSrxSpeedControllerSim.getMotorControllerWrapper(aCanPort);
    }

    private CtrePigeonImuSim getPigeonWrapper(int aCanPort)
    {
        return mPigeonMap.get(aCanPort);
    }

    @SuppressWarnings("PMD.ExcessiveMethodLength")
    public void handleMotorControllerMessage(String aCallback, int aCanPort, ByteBuffer aData)
    {
        aData.order(ByteOrder.LITTLE_ENDIAN);

        sLOGGER.log(Level.INFO, "Handling motor controller message " + aCallback + ", " + aCanPort);

        Set<String> unsupportedFunctions = new HashSet<>();

        if ("Create".equals(aCallback))
        {
            createMotorController(aCanPort);
        }
        else if (mMotorControllerNormalCallbacks.containsKey(aCallback))
        {
            CtreTalonSrxSpeedControllerSim wrapper = getMotorControllerWrapper(aCanPort);
            mMotorControllerNormalCallbacks.get(aCallback).accept(wrapper);
        }
        else if (mMotorControllerSlottedCallbacks.containsKey(aCallback))
        {
            CtreTalonSrxSpeedControllerSim wrapper = getMotorControllerWrapper(aCanPort);

            int slot = aData.getInt();
            mMotorControllerSlottedCallbacks.get(aCallback).accept(wrapper, slot);
        }
        ///////////////////////////////////
        // Unsupported, but not important
        ///////////////////////////////////
        else if (unsupportedFunctions.contains(aCallback))
        {
            sLOGGER.log(Level.DEBUG, aCallback + " not supported");
        }
        else
        {
            sLOGGER.log(Level.ERROR, "Unknown motor callback: " + aCallback);
        }
    }

    public CtrePigeonImuSim createPigeon(int aPort)
    {
        if (mPigeonMap.containsKey(aPort))
        {
            sLOGGER.log(Level.WARN, "Pigeon already registerd on " + aPort);
            return null;
        }
        else
        {
            CtrePigeonImuSim sim = new CtrePigeonImuSim(aPort, CtrePigeonImuSim.sCTRE_OFFSET + aPort * 3);
            mPigeonMap.put(aPort, sim);
            return sim;
        }

    }

    public void createMotorController(int aCanPort)
    {
        int simPort = aCanPort + BaseCanSmartSpeedController.sCAN_SC_OFFSET;
        if (!DataAccessorFactory.getInstance().getSpeedControllerAccessor().getPortList().contains(simPort))
        {
            sLOGGER.log(Level.WARN, "CTRE Motor Controller is being created dynamically instead of in the config file for port " + aCanPort);

            DataAccessorFactory.getInstance().getSpeedControllerAccessor().createSimulator(simPort, CtreTalonSrxSpeedControllerSim.class.getName());
        }
        else if (!(SensorActuatorRegistry.get().getSpeedControllers().get(simPort) instanceof CtreTalonSrxSpeedControllerSim))
        {
            sLOGGER.log(Level.FATAL, "A pre-registered motor controller of type " + SensorActuatorRegistry.get().getSpeedControllers().get(simPort).getClass()
                + " is the wrong type on port " + aCanPort);
            SensorActuatorRegistry.get().getSpeedControllers().remove(simPort);
            createMotorController(aCanPort);
            return;
        }
        SensorActuatorRegistry.get().getSpeedControllers().get(aCanPort + CtreTalonSrxSpeedControllerSim.sCAN_SC_OFFSET).setInitialized(true);
    }

    public void handlePigeonMessage(String aName, int aPort, ByteBuffer aData)
    {
        aData.order(ByteOrder.LITTLE_ENDIAN);

        sLOGGER.log(Level.TRACE, "Handling pigeon message " + aName + ", " + aPort);

        if ("Create".equals(aName))
        {
            CtrePigeonImuSim pigeon = mPigeonMap.get(aPort);
            if (pigeon == null)
            {
                sLOGGER.log(Level.WARN, "CTRE Pigeon is being created dynamically instead of in the config file for port " + aPort);
                pigeon = createPigeon(aPort);
            }

            pigeon.setInitialized(true);
        }
        else if (mPigeonNormalCallbacks.containsKey(aName))
        {
            CtrePigeonImuSim wrapper = getPigeonWrapper(aPort);
            mPigeonNormalCallbacks.get(aName).accept(wrapper);
        }
        else
        {
            sLOGGER.log(Level.ERROR, "Unsupported option " + aName + "(" + aData.limit() + " bytes)");
        }
    }

    public void handleCanifierMessage(String aName, int aDeviceId, ByteBuffer aBuffer)
    {
        sLOGGER.log(Level.ERROR, "Unknown canifier callback: " + aName);
    }

    public void handleBuffTrajPointStreamMessage(String aName, int aDeviceId, ByteBuffer aBuffer)
    {
        sLOGGER.log(Level.ERROR, "Unknown buff traj callback: " + aName);
    }
}

File: snobot_sim/src/main/native/cpp/SnobotSim/SimulatorComponents/CtreWrappers/CtreManager.cpp

#include "SnobotSim/SimulatorComponents/CtreWrappers/CtreManager.h"

#include <cstring>

#include "SnobotSim/GetSensorActuatorHelper.h"
#include "SnobotSim/Logging/SnobotLogger.h"
#include "SnobotSim/ModuleWrapper/Factories/FactoryContainer.h"
#include "SnobotSim/SimulatorComponents/CtreWrappers/CtrePigeonImuSim.h"
#include "SnobotSim/SimulatorComponents/CtreWrappers/CtreTalonSRXSpeedControllerSim.h"
#include "SnobotSim/SimulatorComponents/SmartSC/CanIdOffset.h"
#include "SnobotSim/SimulatorComponents/SmartSC/UnpackingUtils.h"

using namespace SnobotSim;

CtreManager::CtreManager()
{
    mMotorControllerNormalCallbacks["Set_4"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleSet4(); };
    mMotorControllerNormalCallbacks["ConfigSelectedFeedbackSensor"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleConfigSelectedFeedbackSensor(); };
    mMotorControllerNormalCallbacks["ConfigMotionCruiseVelocity"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleSetMotionCruiseVelocity(); };
    mMotorControllerNormalCallbacks["ConfigMotionAcceleration"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleSetMotionAcceleration(); };
    mMotorControllerNormalCallbacks["PushMotionProfileTrajectory"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handlePushMotionProfileTrajectory(); };
    mMotorControllerNormalCallbacks["PushMotionProfileTrajectory_2"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handlePushMotionProfileTrajectory_2(); };
    mMotorControllerNormalCallbacks["SetSelectedSensorPosition"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleSetSelectedSensorPosition(); };
    mMotorControllerNormalCallbacks["SetInverted_2"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleSetInverted_2(); };

    mMotorControllerNormalCallbacks["GetMotorOutputPercent"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetMotorOutputPercent(); };
    mMotorControllerNormalCallbacks["GetSelectedSensorPosition"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetSelectedSensorPosition(); };
    mMotorControllerNormalCallbacks["GetSelectedSensorVelocity"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetSelectedSensorVelocity(); };
    mMotorControllerNormalCallbacks["GetClosedLoopError"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetClosedLoopError(); };
    mMotorControllerNormalCallbacks["GetMotionProfileStatus"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetMotionProfileStatus(); };
    mMotorControllerNormalCallbacks["GetActiveTrajectoryPosition"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetActiveTrajectoryPosition(); };
    mMotorControllerNormalCallbacks["GetActiveTrajectoryVelocity"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetActiveTrajectoryVelocity(); };
    mMotorControllerNormalCallbacks["GetQuadraturePosition"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetQuadraturePosition(); };
    mMotorControllerNormalCallbacks["GetQuadratureVelocity"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper) { wrapper->handleGetQuadratureVelocity(); };

    mMotorControllerSlottedCallbacks["Config_kP"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper, int slot) { wrapper->handleSetKP(slot); };
    mMotorControllerSlottedCallbacks["Config_kI"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper, int slot) { wrapper->handleSetKI(slot); };
    mMotorControllerSlottedCallbacks["Config_kD"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper, int slot) { wrapper->handleSetKD(slot); };
    mMotorControllerSlottedCallbacks["Config_kF"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper, int slot) { wrapper->handleSetKF(slot); };
    mMotorControllerSlottedCallbacks["Config_IntegralZone"] = [](std::shared_ptr<CtreTalonSRXSpeedControllerSim> wrapper, int slot) { wrapper->handleSetIntegralZone(slot); };

    mPigeonNormalCallbacks["SetYaw"] = [](std::shared_ptr<CtrePigeonImuSim> wrapper) { wrapper->handleSetYaw(); };
    mPigeonNormalCallbacks["GetRawGyro"] = [](std::shared_ptr<CtrePigeonImuSim> wrapper) { wrapper->handleGetRawGyro(); };
    mPigeonNormalCallbacks["GetYawPitchRoll"] = [](std::shared_ptr<CtrePigeonImuSim> wrapper) { wrapper->handleGetYawPitchRoll(); };
    mPigeonNormalCallbacks["GetFusedHeading"] = [](std::shared_ptr<CtrePigeonImuSim> wrapper) { wrapper->handleGetFusedHeading(); };
    mPigeonNormalCallbacks["GetFusedHeading1"] = [](std::shared_ptr<CtrePigeonImuSim> wrapper) { wrapper->handleGetFusedHeading(); };
}

void CtreManager::Reset()
{
    mPigeonMap.clear();
}

void CtreManager::handleMotorControllerMessage(const std::string& aCallback, int aCanPort, uint8_t* aBuffer, int aLength)
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_DEBUG, "Getting Motor Controller Message " << aCallback << " on port " << aCanPort << "(" << aLength << " bytes)");

    if ("Create" == aCallback)
    {
        createMotorController(aCanPort);
    }
    else if (mMotorControllerNormalCallbacks.find(aCallback) != mMotorControllerNormalCallbacks.end())
    {
        auto wrapper = CtreTalonSRXSpeedControllerSim::getMotorControllerWrapper(aCanPort);
        mMotorControllerNormalCallbacks[aCallback](wrapper);
    }
    else if (mMotorControllerSlottedCallbacks.find(aCallback) != mMotorControllerSlottedCallbacks.end())
    {
        auto wrapper = CtreTalonSRXSpeedControllerSim::getMotorControllerWrapper(aCanPort);
        auto [slot, value] = ExtractData<int, double>(aBuffer, aLength);

        mMotorControllerSlottedCallbacks[aCallback](wrapper, slot);
    }
    else
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Unknown motor callback: " << aCallback);
    }
}

void CtreManager::handlePigeonMessage(const std::string& aCallback, int aCanPort, uint8_t* aBuffer, int aLength)
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_DEBUG, "Getting Pigeon Message " << aCallback << " on port " << aCanPort << "(" << aLength << " bytes)");

    if ("Create" == aCallback)
    {
        if (mPigeonMap.find(aCanPort) == mPigeonMap.end())
        {
            SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "CTRE Pigeon is being created dynamically instead of in the config file for port " << aCanPort);
            createPigeon(aCanPort);
        }

        mPigeonMap[aCanPort]->SetInitialized(true);
    }
    else if (mPigeonNormalCallbacks.find(aCallback) != mPigeonNormalCallbacks.end())
    {
        auto wrapper = getPigeonWrapper(aCanPort);
        mPigeonNormalCallbacks[aCallback](wrapper);
    }
    else
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Unknown pigeon callback: " << aCallback);
    }
}

void CtreManager::handleCanifierMessage(const std::string& aCallback, int aCanPort, uint8_t* aBuffer, int aLength)
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Getting Canifier Message " << aCallback << " on port " << aCanPort << "(" << aLength << " bytes)");
}

void CtreManager::handleCanCoderMessage(const std::string& aCallback, int aCanPort, uint8_t* aBuffer, int aLength)
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Getting CanCoder Message " << aCallback << " on port " << aCanPort << "(" << aLength << " bytes)");
}

void CtreManager::handleBuffTrajPointStreamMessage(const std::string& aCallback, int aCanPort, uint8_t* aBuffer, int aLength)
{
    SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Getting Trajectory Point Message " << aCallback << " on port " << aCanPort << "(" << aLength << " bytes)");
}

std::shared_ptr<CtrePigeonImuSim> CtreManager::createPigeon(int aPort)
{
    if (mPigeonMap.find(aPort) != mPigeonMap.end())
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Pigeon alrady registered on " << aPort);
    }
    else
    {
        std::shared_ptr<CtrePigeonImuSim> sim(new CtrePigeonImuSim(aPort, CtrePigeonImuSim::CTRE_OFFSET + aPort * 3));
        mPigeonMap[aPort] = sim;
    }

    return mPigeonMap[aPort];
}

void CtreManager::createMotorController(int aCanPort)
{
    int simPort = aCanPort + CAN_SC_OFFSET;

    auto wrapper = SensorActuatorRegistry::Get().GetISpeedControllerWrapper(aCanPort, false);
    if (!wrapper)
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_WARN, "CTRE Motor Controller is being created dynamically instead of in the config file for port " << aCanPort);

        FactoryContainer::Get().GetSpeedControllerFactory()->Create(simPort, CtreTalonSRXSpeedControllerSim::TYPE);
        // SensorActuatorRegistry::Get().Register(simPort, std::shared_ptr<ISpeedControllerWrapper>(new CtreTalonSRXSpeedControllerSim(aCanPort)));
    }
    else if (!std::dynamic_pointer_cast<CtreTalonSRXSpeedControllerSim>(wrapper))
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "Wrong motor type set up on " << aCanPort);
        return;
    }

    SensorActuatorRegistry::Get().GetISpeedControllerWrapper(simPort, true)->SetInitialized(true);
}

std::shared_ptr<CtrePigeonImuSim> CtreManager::getPigeonWrapper(int aCanPort)
{
    if (mPigeonMap.find(aCanPort) == mPigeonMap.end())
    {
        SNOBOT_LOG(SnobotLogging::LOG_LEVEL_CRITICAL, "No pigeon on " << aCanPort);
        StackTraceHelper::PrintStackTracker();
    }
    return mPigeonMap[aCanPort];
}
